rust/src/http2/parser.rs

```diff
@@ -0,0 +1,1718 @@
+    match r {
+        Ok((r, head)) => {
+            let headcopy = HTTP2FrameHeaderBlock {
+                name: head.name.to_vec(),
+                value: head.value.to_vec(),
+                error: head.error,
+            };
+            dyn_headers.push(headcopy);
+            if dyn_headers.len() > 255 - HTTP2_STATIC_HEADERS_NUMBER {
+                dyn_headers.remove(0);
+            }
+            //TODOnext? handle dynamic table size limit
+            return Ok((r, head));
+        }
+        Err(e) => {
+            return Err(e);
+        }
+    }
+}
+
+fn http2_parse_headers_block_literal_noindex<'a>(
+    input: &'a [u8],
+    dyn_headers: &Vec<HTTP2FrameHeaderBlock>,
+) -> IResult<&'a [u8], HTTP2FrameHeaderBlock> {
+    fn parser(input: &[u8]) -> IResult<&[u8], (u8, u8)> {
+        bits!(
+            input,
+            complete!(tuple!(
+                verify!(take_bits!(4u8), |&x| x == 0),
+                take_bits!(4u8)
+            ))
+        )
+    }
+    let (i2, indexed) = parser(input)?;
+    let r = http2_parse_headers_block_literal_common(i2, indexed.1, dyn_headers);
+    return r;
+}
+
+fn http2_parse_headers_block_literal_neverindex<'a>(
+    input: &'a [u8],
+    dyn_headers: &Vec<HTTP2FrameHeaderBlock>,
+) -> IResult<&'a [u8], HTTP2FrameHeaderBlock> {
+    fn parser(input: &[u8]) -> IResult<&[u8], (u8, u8)> {
+        bits!(
+            input,
+            complete!(tuple!(
+                verify!(take_bits!(4u8), |&x| x == 1),
+                take_bits!(4u8)
+            ))
+        )
+    }
+    let (i2, indexed) = parser(input)?;
+    let r = http2_parse_headers_block_literal_common(i2, indexed.1, dyn_headers);
+    return r;
+}
+
+fn http2_parse_headers_block_dynamic_size(input: &[u8]) -> IResult<&[u8], HTTP2FrameHeaderBlock> {
+    fn parser(input: &[u8]) -> IResult<&[u8], (u8, u8)> {
+        bits!(
+            input,
+            complete!(tuple!(
+                verify!(take_bits!(3u8), |&x| x == 1),
+                take_bits!(5u8)
+            ))
+        )
+    }
+    let (i2, maxsize) = parser(input)?;
+    if maxsize.1 == 31 {
+        let (i3, _maxsize2) = take_while!(i2, |ch| (ch & 0x80) != 0)?;
+        let (i4, _maxsize3) = take!(i3, 1)?;
+        //TODOnext detect on Dynamic Table Size Update RFC6.3
+        return Ok((
+            i4,
+            HTTP2FrameHeaderBlock {
+                name: Vec::new(),
+                value: Vec::new(),
+                error: HTTP2HeaderDecodeStatus::HTTP2HeaderDecodeSizeUpdate,
+            },
+        ));
+    }
+    return Ok((
+        i2,
+        HTTP2FrameHeaderBlock {
+            name: Vec::new(),
+            value: Vec::new(),
+            error: HTTP2HeaderDecodeStatus::HTTP2HeaderDecodeSizeUpdate,
+        },
+    ));
+}
+
+fn http2_parse_headers_block<'a>(
+    input: &'a [u8],
+    dyn_headers: &mut Vec<HTTP2FrameHeaderBlock>,
+) -> IResult<&'a [u8], HTTP2FrameHeaderBlock> {
+    //caller garantees o have at least one byte
```

Comment:
We now have a debug_validate in master which might be useful here.  However, it does conflict with unit tests.

Reply:
It is not yet in master, is it ?
`grep -r debug_validate rust/src/` yields no result

Reply:
cf https://github.com/OISF/suricata/pull/4969